File: protocol/clock.js
/* What a song clock has to do, and a test that a guessing clock cannot pass.
   ---------------------------------------------------------------------------
   The protocol answers "where are we in the song" in musical terms, but
   something has to tell it where the song is in seconds. That something is a
   clock, and everything downstream is only as good as the clock is honest.

   A consumer can get this wrong in two different ways, and they are not the
   same mistake.

     COUNTING     Start a timer when play is pressed and report elapsed time --
                  from performance.now(), or worse, by adding up frame deltas.
                  It starts before the sound does, so it runs ahead; the size of
                  that head start changes every run, so no fixed calibration
                  removes it; and it drifts from the audio over a song. This is
                  accurate never.

     READING RAW  Return audio.currentTime directly. Accurate, but an audio
                  element only updates that value a few times a second, so the
                  reading is a staircase. Anything drawing from it judders: in a
                  game the notes visibly step toward you instead of sliding.
                  This is accurate but not smooth.

     ANCHORED     Read audio.currentTime, and between its updates carry the
                  position forward with a wall clock -- re-anchoring every time
                  the audio reports a new value. Accurate AND smooth, because
                  the wall clock is only ever trusted for the few milliseconds
                  since the last real measurement, so its error can never build.

   Hand one to a Session as `songTime` and the protocol stops keeping a clock
   of its own:

       const clock = AnchoredClock(audio);
       const s = Session(score, { songTime: () => clock.position() });

   conformance() is what makes this a rule rather than advice. It drives a
   clock through an audio element that starts late, runs slightly fast, and
   updates coarsely, and reports where the clock lies. All three clocks above
   are in this file on purpose: a test that nothing can fail is not a test, so
   the suite proves the harness by failing the two that should fail.
*/
"use strict";

/* ---- a clock that counts. Kept so the test has something to fail. -------- */
function CountingClock(audio, opts) {
  const now = (opts && opts.now) || (() => performance.now() / 1000);
  let anchor = 0, anchorWall = null, playing = false;
  return {
    get playing() { return playing; },
    position() {
      if (!playing || anchorWall === null) return anchor;
      return anchor + (now() - anchorWall);
    },
    play(at) {
      anchor = at === undefined ? this.position() : Math.max(0, at);
      anchorWall = now();                 /* counting starts here... */
      playing = true;
      audio.play(anchor);                 /* ...and the sound starts later */
    },
    pause() { anchor = this.position(); playing = false; audio.pause(); },
    seek(t) { anchor = Math.max(0, t); anchorWall = now(); if (playing) audio.play(anchor); },
  };
}

/* ---- a clock that reads the audio raw. Accurate, and it judders. --------- */
function RawAudioClock(audio) {
  return {
    get playing() { return !audio.paused; },
    position() { return audio.currentTime; },
    play(at) { if (at !== undefined) audio.currentTime = Math.max(0, at); audio.play(); },
    pause() { audio.pause(); },
    seek(t) { audio.currentTime = Math.max(0, t); },
  };
}

/* ---- the one to use ----------------------------------------------------- */
function AnchoredClock(audio, opts) {
  const now = (opts && opts.now) || (() => performance.now() / 1000);
  /* How far the wall clock is allowed to carry the position on its own before
     we stop believing it. An audio element updates a few times a second; if it
     has told us nothing for this long, something is wrong and a frozen reading
     is a better answer than an invented one. */
  const maxCarry = (opts && opts.max_carry_s) || 0.25;

  let lastRaw = null;      /* the last value the audio actually reported */
  let rawAt = 0;           /* wall time when it reported it */
  let out = 0;             /* what we last returned, so we never go backwards */

  function position() {
    const raw = audio.currentTime;
    if (audio.paused) { lastRaw = raw; rawAt = now(); out = raw; return raw; }

    if (raw !== lastRaw) {               /* a real measurement arrived */
      lastRaw = raw;
      rawAt = now();
    }
    const carried = now() - rawAt;
    /* Between the audio's updates, carry forward with the wall clock. The error
       can never accumulate, because the next real reading resets it. */
    let est = lastRaw + Math.min(carried, maxCarry);
    /* Time in a song does not go backwards, and a note that jumps backwards is
       a missed hit the player did not earn. */
    if (est < out) est = out;
    out = est;
    return est;
  }

  return {
    get playing() { return !audio.paused; },
    position,
    play(at) {
      if (at !== undefined) audio.currentTime = Math.max(0, at);
      lastRaw = null; out = audio.currentTime; rawAt = now();
      audio.play();
    },
    pause() { audio.pause(); lastRaw = null; out = audio.currentTime; },
    seek(t) {
      audio.currentTime = Math.max(0, t);
      lastRaw = null; out = audio.currentTime; rawAt = now();
    },
  };
}

/* ---- an audio element that misbehaves the way real ones do --------------- */
/* start_delay_s  the gap between play() and the first sample being audible.
   update_every_s how often currentTime actually changes -- an element reports
                  a few times a second, not continuously.
   rate           the device's sample clock against the CPU's. 1.0003 means the
                  card plays slightly fast, which is an ordinary amount wrong. */
function FakeAudio(now, opts) {
  opts = opts || {};
  const startDelay = opts.start_delay_s === undefined ? 0.08 : opts.start_delay_s;
  const updateEvery = opts.update_every_s === undefined ? 0.05 : opts.update_every_s;
  const rate = opts.rate === undefined ? 1 : opts.rate;
  let from = 0, startedWall = null, reported = 0;

  const self = {
    paused: true,
    get currentTime() {
      if (startedWall === null) return reported;
      const t = self.trueTime();
      /* only report on the update grid: this is the staircase */
      const steps = Math.floor((t - from) / updateEvery);
      reported = steps < 0 ? from : from + steps * updateEvery;
      return reported;
    },
    set currentTime(v) { from = Math.max(0, v); reported = from; startedWall = null; },
    trueTime() {
      if (startedWall === null) return from;
      const e = now() - startedWall - startDelay;
      return e < 0 ? from : from + e * rate;
    },
    play(at) {
      if (at !== undefined) { from = Math.max(0, at); reported = from; }
      startedWall = now();
      self.paused = false;
    },
    pause() { from = self.trueTime(); reported = from; startedWall = null; self.paused = true; },
  };
  return self;
}

function conformance(makeClock, label) {
  const res = [];
  const ok = (n, c, d) => res.push([!!c, (label ? label + ": " : "") + n, d || ""]);

  /* --- 1. the sound starts late, and the clock must not run ahead of it --- */
  let t = 0; const now = () => t;
  let a = FakeAudio(now, { start_delay_s: 0.080, update_every_s: 0.001, rate: 1 });
  let c = makeClock(a, { now });
  c.play(0);
  t += 0.040;                                  /* 40 ms in: still silent */
  const silent = (c.position() - a.trueTime()) * 1000;
  ok("does not run ahead of a sound that has not started", Math.abs(silent) < 8,
     `${silent >= 0 ? "+" : ""}${silent.toFixed(0)} ms ahead while still silent`);

  t += 1.0;
  const started = (c.position() - a.trueTime()) * 1000;
  ok("agrees with the audio once it is playing", Math.abs(started) < 8,
     `${started >= 0 ? "+" : ""}${started.toFixed(0)} ms`);

  /* --- 2. the same test with a different start delay ---------------------- */
  /* The one a fixed calibration cannot survive: if the error differs between
     these two, no single number corrects both. */
  t = 0;
  let a2 = FakeAudio(now, { start_delay_s: 0.020, update_every_s: 0.001, rate: 1 });
  let c2 = makeClock(a2, { now });
  c2.play(0);
  t += 1.020;
  const other = (c2.position() - a2.trueTime()) * 1000;
  ok("is wrong by the same amount when the start delay changes",
     Math.abs(other - started) < 8,
     `${started.toFixed(0)} ms at one delay, ${other.toFixed(0)} ms at another`
     + (Math.abs(other - started) >= 8 ? " -- a fixed calibration cannot fix both" : ""));

  /* --- 3. the card runs fast, and the clock must run fast with it --------- */
  t = 0;
  let a3 = FakeAudio(now, { start_delay_s: 0, update_every_s: 0.001, rate: 1.0003 });
  let c3 = makeClock(a3, { now });
  c3.play(0);
  t += 1.0; const early = (c3.position() - a3.trueTime()) * 1000;
  t += 240.0; const late = (c3.position() - a3.trueTime()) * 1000;
  ok("does not slide apart from the audio across a song", Math.abs(late - early) < 50,
     `${early.toFixed(0)} ms at the start, ${late.toFixed(0)} ms four minutes in`);

  /* --- 4. smooth enough to draw from ------------------------------------- */
  /* A game reads the clock once per frame and moves everything by the
     difference. If the clock only changes a few times a second, the reading is
     a staircase and the notes step toward the player instead of sliding. */
  t = 0;
  let a4 = FakeAudio(now, { start_delay_s: 0, update_every_s: 0.050, rate: 1 });
  let c4 = makeClock(a4, { now });
  c4.play(0);
  const frame = 1 / 60;
  let prev = c4.position(), biggest = 0, stalls = 0;
  for (let i = 0; i < 120; i++) {
    t += frame;
    const p = c4.position();
    const step = p - prev;
    if (step > biggest) biggest = step;
    if (step < frame * 0.25) stalls++;
    prev = p;
  }
  ok("moves by about one frame per frame, so nothing judders",
     biggest < frame * 2.5 && stalls < 12,
     `biggest jump ${(biggest * 1000).toFixed(0)} ms (a frame is ${(frame * 1000).toFixed(0)}), `
     + `${stalls} of 120 frames stalled`);

  /* --- 5. never backwards ------------------------------------------------ */
  t = 0;
  let a5 = FakeAudio(now, { start_delay_s: 0.03, update_every_s: 0.050, rate: 1 });
  let c5 = makeClock(a5, { now });
  c5.play(0);
  let last = -Infinity, back = 0;
  for (let i = 0; i < 200; i++) { t += frame / 2; const p = c5.position(); if (p < last - 1e-9) back++; last = p; }
  ok("never goes backwards", back === 0, `${back} backward steps`);

  /* --- 6. paused means paused, and a seek lands where it was sent --------- */
  t = 0;
  let a6 = FakeAudio(now, { start_delay_s: 0, update_every_s: 0.001, rate: 1 });
  let c6 = makeClock(a6, { now });
  c6.play(10);
  t += 2.0;
  const held = c6.position();
  c6.pause();
  t += 30.0;
  ok("does not move while paused", Math.abs(c6.position() - held) * 1000 < 8,
     `moved ${((c6.position() - held) * 1000).toFixed(0)} ms in 30 s of pause`);
  c6.seek(61.9);
  ok("reports where it was sent", Math.abs(c6.position() - 61.9) < 0.008,
     `${c6.position().toFixed(3)} s`);

  return res;
}

const M = { CountingClock, RawAudioClock, AnchoredClock, FakeAudio, conformance };
if (typeof module !== "undefined" && module.exports) module.exports = M;
if (typeof window !== "undefined") window.LimelightClock = M;

File: protocol/clock.py
"""What a song clock has to do, and a test that a guessing clock cannot pass.

The protocol answers "where are we in the song" in musical terms, but something
has to tell it where the song is in seconds. That something is a clock, and a
show is only as well synchronised as its clock is honest.

There are two ways to build one.

  COUNTING   Note the wall time when play is pressed, then report the elapsed
             wall seconds since. Cheap, needs nothing from the audio, and wrong
             in three ways: it starts counting before the sound starts, so it
             runs ahead by however long the audio took to begin; that head start
             is a different number every time, so no fixed delay setting can
             remove it; and a sound card's sample clock does not run at exactly
             the speed the CPU thinks it does, so the two slide apart over a
             song.

  MEASURING  Ask the audio device how much sound it has actually delivered to
             the speaker. Nothing to estimate: if the device started late, the
             number says so; if the card runs slow, the number runs slow with
             it. A delay setting survives, but it goes back to meaning the only
             thing it should ever have meant -- how long the lamps and the cable
             take to react, which really is the same every time.

The contract is one method:

    position() -> float     song seconds, measured, honest while paused

Hand it to a Session as `song_time` and the protocol reads it instead of
running a clock of its own:

    clock = MeasuredClock(device)
    s = Session(score, song_time=clock.position)

`conformance()` below is the part that makes this a rule rather than advice.
It drives a clock through a device that starts late and runs at the wrong
speed, and fails anything that answers with arithmetic instead of measurement.
CountingClock is kept in this file on purpose: a test that nothing can fail is
not a test, so the suite proves the harness by failing that one.
"""
import math


# ---------------------------------------------------------------------------
# The device a measuring clock needs.
#
# Three methods. An implementation over sounddevice is about fifteen lines:
# open an OutputStream, count the frames handed to the callback, and report
# frames / samplerate minus the stream's own output latency -- because frames
# handed to the driver have not reached the speaker yet.
# ---------------------------------------------------------------------------
class Device:
    def start(self, at: float) -> None:
        """Begin playing from `at` song seconds."""
        raise NotImplementedError

    def stop(self) -> None:
        raise NotImplementedError

    def played(self):
        """Song seconds currently audible, or None if no sound is out yet.

        None is not zero and the difference matters: between pressing play and
        the first sample reaching the speaker, a clock that says 0.0 is making
        the same claim as one that has measured silence.
        """
        raise NotImplementedError


class MeasuredClock:
    """A clock that asks the device where the music is."""

    def __init__(self, device: Device, latency_s: float = 0.0):
        self.device = device
        self.latency_s = float(latency_s)   # lamps and cable, not audio startup
        self._anchor = 0.0                  # where we were told to be
        self._playing = False

    @property
    def playing(self) -> bool:
        return self._playing

    def position(self) -> float:
        if not self._playing:
            return self._anchor
        p = self.device.played()
        # Nothing audible yet: hold at the anchor rather than run ahead of the
        # sound. This is the head start that a counting clock builds in and
        # then spends an evening trying to tune out.
        if p is None:
            return self._anchor
        return p + self.latency_s

    def play(self, at=None) -> None:
        if at is None:
            at = self.position()
        self._anchor = max(0.0, float(at))
        self._playing = True
        self.device.start(self._anchor)

    def pause(self) -> None:
        self._anchor = self.position()
        self._playing = False
        self.device.stop()

    def seek(self, to: float) -> None:
        was = self._playing
        self.device.stop()
        self._anchor = max(0.0, float(to))
        if was:
            self.device.start(self._anchor)
        self._playing = was


class CountingClock:
    """The clock this project has been using, kept so the test can fail it.

    It never asks the device anything. Every number it returns is arithmetic on
    wall time, which is why a show driven by it needs a delay setting, why the
    right value for that setting changes between runs, and why the lights are
    furthest out at the end of the song.
    """

    def __init__(self, device: Device, now, latency_s: float = 0.0):
        self.device = device
        self.now = now
        self.latency_s = float(latency_s)
        self._anchor = 0.0
        self._anchor_wall = None
        self._playing = False

    @property
    def playing(self) -> bool:
        return self._playing

    def position(self) -> float:
        if not self._playing or self._anchor_wall is None:
            return self._anchor
        return self._anchor + (self.now() - self._anchor_wall) + self.latency_s

    def play(self, at=None) -> None:
        if at is None:
            at = self.position()
        self._anchor = max(0.0, float(at))
        self._anchor_wall = self.now()      # counting starts here...
        self._playing = True
        self.device.start(self._anchor)     # ...and the sound starts later

    def pause(self) -> None:
        self._anchor = self.position()
        self._playing = False
        self.device.stop()

    def seek(self, to: float) -> None:
        was = self._playing
        self.device.stop()
        self._anchor = max(0.0, float(to))
        self._anchor_wall = self.now()
        if was:
            self.device.start(self._anchor)
        self._playing = was


# ---------------------------------------------------------------------------
# A device that misbehaves the way real ones do, so a clock can be tested
# without a sound card.
# ---------------------------------------------------------------------------
class FakeDevice(Device):
    """Starts late and runs at the wrong speed, both by a stated amount.

    `start_delay_s` is the gap between being told to play and the first sample
    reaching the speaker -- process spawn, buffer fill, and on the panel today
    a temporary copy of the rest of the track being written to disk, which is
    why it is not the same number twice.

    `rate` is the device's sample clock measured against the CPU's: 1.0003 means
    the card plays slightly fast, which is an ordinary amount of wrong.
    """

    def __init__(self, now, start_delay_s=0.08, rate=1.0):
        self.now = now
        self.start_delay_s = float(start_delay_s)
        self.rate = float(rate)
        self._from = None
        self._started_wall = None

    def start(self, at: float) -> None:
        self._from = float(at)
        self._started_wall = self.now()

    def stop(self) -> None:
        self._from = None
        self._started_wall = None

    def played(self):
        if self._from is None:
            return None
        elapsed = self.now() - self._started_wall - self.start_delay_s
        if elapsed < 0:
            return None                      # told to play; nothing audible yet
        return self._from + elapsed * self.rate

    def true_position(self):
        """What is actually coming out of the speaker. The thing being checked."""
        p = self.played()
        return self._from if p is None else p


def conformance(make_clock, label=""):
    """Drive a clock through a misbehaving device and report where it lies.

    Returns a list of (passed, name, detail). Tolerances are in milliseconds and
    chosen for what an audience can see: a fixed 20 ms error is invisible, a
    varying one is not, and 50 ms of slide across a song is a drummer playing
    to a different click.
    """
    res = []
    def ok(name, cond, detail=""):
        res.append((bool(cond), (label + ": " if label else "") + name, detail))

    # --- 1. the sound starts late, and the clock must not run ahead of it ----
    t = [0.0]
    now = lambda: t[0]
    dev = FakeDevice(now, start_delay_s=0.080, rate=1.0)
    c = make_clock(dev, now)
    c.play(0.0)
    t[0] += 0.040                       # 40 ms in: still silent
    err_silent = (c.position() - dev.true_position()) * 1000
    ok("does not run ahead of a sound that has not started",
       abs(err_silent) < 5, f"{err_silent:+.0f} ms ahead while still silent")

    t[0] += 1.000                       # 1.04 s of wall, 0.96 s of music
    err_start = (c.position() - dev.true_position()) * 1000
    ok("agrees with the audio once it is playing",
       abs(err_start) < 5, f"{err_start:+.0f} ms")

    # --- 2. the same test, with a different start delay ---------------------
    # This is the one a fixed delay setting cannot survive. If the error here
    # differs from the error above, no single number can correct both.
    t2 = [0.0]
    now2 = lambda: t2[0]
    dev2 = FakeDevice(now2, start_delay_s=0.020, rate=1.0)
    c2 = make_clock(dev2, now2)
    c2.play(0.0)
    t2[0] += 1.040
    err_other = (c2.position() - dev2.true_position()) * 1000
    ok("is wrong by the same amount when the start delay changes",
       abs(err_other - err_start) < 5,
       f"{err_start:+.0f} ms at one delay, {err_other:+.0f} ms at another "
       f"-- a fixed offset cannot fix both" if abs(err_other - err_start) >= 5 else
       f"{err_start:+.0f} vs {err_other:+.0f} ms")

    # --- 3. the card runs fast, and the clock must run fast with it ---------
    t3 = [0.0]
    now3 = lambda: t3[0]
    dev3 = FakeDevice(now3, start_delay_s=0.0, rate=1.0003)
    c3 = make_clock(dev3, now3)
    c3.play(0.0)
    t3[0] += 1.0
    early = (c3.position() - dev3.true_position()) * 1000
    t3[0] += 240.0                      # a four-minute song
    late = (c3.position() - dev3.true_position()) * 1000
    ok("does not slide apart from the audio across a song",
       abs(late - early) < 50, f"{early:+.0f} ms at the start, {late:+.0f} ms four minutes in")

    # --- 4. paused means paused, and a seek lands where it was sent ---------
    t4 = [0.0]
    now4 = lambda: t4[0]
    dev4 = FakeDevice(now4, start_delay_s=0.0, rate=1.0)
    c4 = make_clock(dev4, now4)
    c4.play(10.0)
    t4[0] += 2.0
    held = c4.position()
    c4.pause()
    t4[0] += 30.0
    ok("does not move while paused", abs(c4.position() - held) * 1000 < 5,
       f"moved {(c4.position() - held) * 1000:+.0f} ms in 30 s of pause")
    c4.seek(61.9)
    ok("reports where it was sent", abs(c4.position() - 61.9) < 0.005,
       f"{c4.position():.3f} s")

    return res


# ---------------------------------------------------------------------------
# A real device, over sounddevice. Written against the library's documented
# behaviour and checked only by the conformance harness above, which uses a
# simulated device -- it has not yet been run against a sound card. Treat the
# first run on the rig as the real test.
# ---------------------------------------------------------------------------
class SoundDeviceOutput(Device):
    """Plays a WAV and reports how much of it has actually reached the speaker.

    The measurement is the whole point and it is two lines: count the frames
    handed to the callback, then subtract the stream's own output latency,
    because frames given to the driver are not audible yet. No estimate of
    startup time is needed or made -- before the first callback runs, played()
    says None, which is the honest answer.

    Replaces pw-play, which cannot report its position at all. That is the only
    reason the panel had to guess.
    """

    def __init__(self, path, blocksize=0):
        import soundfile as sf
        self.path = path
        info = sf.info(path)
        self.sr, self.channels = info.samplerate, info.channels
        self.frames_total = info.frames
        self.blocksize = blocksize
        self._stream = None
        self._file = None
        self._from = 0.0
        self._frames = 0
        self._done = False

    @property
    def duration(self):
        return self.frames_total / self.sr

    def start(self, at: float) -> None:
        import sounddevice as sd
        import soundfile as sf
        self.stop()
        self._from = max(0.0, float(at))
        self._frames = 0
        self._done = False
        self._file = sf.SoundFile(self.path)
        self._file.seek(min(int(round(self._from * self.sr)), self.frames_total))

        def callback(outdata, frames, time_info, status):
            data = self._file.read(frames, dtype="float32", always_2d=True)
            n = len(data)
            outdata[:n] = data
            if n < frames:
                outdata[n:] = 0
                self._done = True
                raise sd.CallbackStop
            self._frames += frames

        self._stream = sd.OutputStream(samplerate=self.sr, channels=self.channels,
                                       blocksize=self.blocksize, callback=callback)
        self._stream.start()

    def stop(self) -> None:
        if self._stream is not None:
            try:
                self._stream.stop()
                self._stream.close()
            except Exception:
                pass
        self._stream = None
        if self._file is not None:
            try:
                self._file.close()
            except Exception:
                pass
        self._file = None
        self._frames = 0

    def played(self):
        if self._stream is None or self._frames == 0:
            return None                     # nothing audible yet, and say so
        latency = getattr(self._stream, "latency", 0.0) or 0.0
        return self._from + self._frames / self.sr - float(latency)

    @property
    def finished(self):
        return self._done
